source/agora/consensus/EnrollmentManager.d

```diff
@@ -278,15 +281,24 @@ public class EnrollmentManager
 
         if (enroll.utxo_key in self_utxos)
         {
-            this.setEnrollmentKey(enroll.utxo_key);
+            if (this.enroll_key == Hash.init)
+            {
+                this.setEnrollmentKey(enroll.utxo_key);
 
-            // consume pre-images
-            this.catchupPreImageCycle();
-            this.cycle.populate(this.key_pair.v, enroll.utxo_key, true);
+                // consume pre-images
+                this.catchupPreImageCycle();
+                this.cycle.populate(this.key_pair.v, enroll.utxo_key, true);
 
-            // save the count of 'populate' of `PreImageCycle`
-            this.setCycleIndex(
-                (this.cycle.nonce * this.cycle.NumberOfCycles) + this.cycle.index);
+                // save the count of 'populate' of `PreImageCycle`
+                this.setCycleIndex(
+                    (this.cycle.nonce * this.cycle.NumberOfCycles)
+                        + this.cycle.index);
+            }
+            else
```

Comment:
If you move it as the first check and `return` you can simplify the code / diff.